File: Retos/Reto #2 - EL PARTIDO DE TENIS [Media]/java/bramenn.java

import java.util.Arrays;
import java.util.HashMap;
import java.util.List;

class Bramenn {

    public static void main(String[] args) {

        List<String> tennisSets = Arrays.asList("P1", "P1", "P2", "P2", "P1", "P2", "P1", "P1");
        // List<String> tennisSets = Arrays.asList("P1", "P2", "P2", "P2", "P1", "P2", "P1", "P1");
        // List<String> tennisSets = Arrays.asList("P1", "P2", "P2", "P2", "P1", "P2", "P1", "P2");

        Player player_1 = new Player("P1");
        Player player_2 = new Player("P2");

        TennisMatch theGame = new TennisMatch(player_1, player_2, tennisSets);

        theGame.runGame();
    }

}

class Player {

    private String name;
    private int score;

    public Player(String name) {
        this.name = name;
        this.score = 0;
    }

    public String getName() {
        return name;
    }

    public int getScore() {
        return score;
    }

    public void setScore(int score) {
        this.score += score;
    }

}

class TennisMatch {

    private Player player_1;
    private Player player_2;
    private Player winner;
    private List<String> tennisSets;

    static HashMap<Integer, String> tennisPoints;
    static {
        tennisPoints = new HashMap<>();
        tennisPoints.put(0, "Love");
        tennisPoints.put(1, "15");
        tennisPoints.put(2, "30");
        tennisPoints.put(3, "40");
    };

    public TennisMatch(Player player_1, Player player_2, List<String> tennisSets) {
        this.player_1 = player_1;
        this.player_2 = player_2;
        this.tennisSets = tennisSets;
    }

    public Player getPlayer_1() {
        return player_1;
    }

    public Player getPlayer_2() {
        return player_2;
    }

    public Player getWinner() {
        return winner;
    }

    public void setWinner(Player winner) {
        this.winner = winner;
    }

    public void runGame() {
        this.tennisNarrator();

    }

    public void tennisNarrator() {

        String advantageText = "Ventaja %s";
        String scoreText = "%s - %s";

        for (int tennisSet = 0; tennisSet < this.tennisSets.size(); tennisSet++) {

            Player player = this.getPlayerByName(this.tennisSets.get(tennisSet));

            String text = "";

            player.setScore(1);

            switch (tennisSet) {
                case 5:
                case 6:
                    player = this.getPlayerWithHighestScore();

                    if (player == null) {
                        text = "Deuce";
                    } else {
                        text = String.format(advantageText, player.getName());
                    }

                    break;

                case 7:
                    this.setWinner(this.getPlayerWithHighestScore());
                    break;
                default:
                    text = String.format(scoreText, tennisPoints.get(player_1.getScore()),
                            tennisPoints.get(player_2.getScore()));
                    break;
            }

            if (!text.isEmpty()) {
                System.out.println(text);
            }
        }

        try {
            System.out.println("Ha ganado el " + this.getWinner().getName());
        } catch (Exception e) {
            System.out.println("Deuce");
        }
    }

    public Player getPlayerWithHighestScore() {
        Player winner = null;
        if (this.player_1.getScore() > this.player_2.getScore()) {
            winner = player_1;
        } else if (this.player_1.getScore() < this.player_2.getScore()) {
            winner = player_2;
        }
        return winner;
    }

    public Player getPlayerByName(String name) {

        if (name == this.getPlayer_1().getName()) {
            return this.getPlayer_1();
        } else {
            return this.getPlayer_2();
        }
    }
}

File: Retos/Reto #2 - EL PARTIDO DE TENIS [Media]/python/bramenn.py

from typing import List

class Player:
    score: int
    name: str

    def __init__(self, name: str) -> None:
        self.name = name
        self.score = 0

class TennisMatch:

    player_1: Player
    player_2: Player

    history_sets: List[str]
    winner: Player

    tennis_dict = {
        0: "Love",
        1: "15",
        2: "30",
        3: "40",
    }

    def __init__(self, player_1: Player, player_2: Player, history_sets: List[str]) -> None:
        self.player_1 = player_1
        self.player_2 = player_2
        self.history_sets = history_sets


    def run_game(self):
        
        counter_set = 0
        for player_name in self.history_sets:

            if player_name == self.player_1.name:
                self.player_1.score += 1
            elif player_name == self.player_2.name:
                self.player_2.score += 1

            counter_set += 1

            self.load_winner()

            if counter_set in [6, 7]:
                if self.check_deuce:
                   print("Deuce")
                else:
                    print(f"Ventaja {self.winner.name}")
            elif counter_set in [8]:
                if self.check_deuce:
                    print("Deuce")
                    continue


                print(f"Ha ganado el {self.winner.name}")
            else:
                print(f"{self.tennis_dict.get(self.player_1.score)} - {self.tennis_dict.get(self.player_2.score)}") 

    @property
    def check_deuce(self) -> bool:
        return True if self.get_high_scoring_player() is None else False
    
    def load_winner(self):
        
        if self.check_deuce:
            self.winner = None
            return
        
        self.winner = self.get_high_scoring_player()
        


    def get_high_scoring_player(self) -> Player:

        if self.player_1.score > self.player_2.score:
            return self.player_1
        elif self.player_1.score < self.player_2.score:
            return self.player_2


def main():
    player_1 = Player(name="P1")
    player_2 = Player(name="P2")

    history_sets = ["P1", "P1", "P2", "P2", "P1", "P2", "P1", "P1"]
    # history_sets = ["P1", "P2", "P2", "P2", "P1", "P2", "P1", "P1"]
    # history_sets = ["P1", "P2", "P2", "P2", "P1", "P2", "P1", "P2"]
    
    tennis_match = TennisMatch(player_1=player_1, player_2=player_2, history_sets=history_sets)

    tennis_match.run_game()

main()